core/src/main/clojure/xtdb/operator/scan.clj

```diff
@@ -398,6 +399,34 @@
                                  system-from
                                  nil))))))))))
 
+(def ^:private neg-or-zero-int? (complement pos-int?))
+
+(defn- iid-selector [^ByteBuffer iid-bb]
+  (reify IRelationSelector
+    (select [_ _allocator rel-rdr _params]
+      (let [ptr (ArrowBufPointer.)
+            iid-rdr (.readerForName rel-rdr "xt$iid")
+            value-count (.valueCount iid-rdr)]
+        ;; lower-bound
+        (loop [left 0 right (dec value-count)]
+          (if (= left right)
+            (if (zero? (util/compare-nio-buffers-unsigned iid-bb (util/arrow-buf-pointer->bb (.getPointer iid-rdr left ptr))))
+              ;; upper bound
```

Comment:
I don't know if it's worth binary searching again for the upper bound - if there are fewer than log2(N) entries per entity on average it'll be quicker to scan from the lower bound